### PR TITLE
Implement Stage instances

### LINKED ISSUE #602: Implement Stage Instances
https://github.com/discord/discord-api-docs/commit/f019fc358047050513c623f3639b6e96809f9280

### PR DESCRIPTION
**Pull request by ashwinvin**
### Summary
Implements Stage Instances

### Checklist
- [x] I have run `nox` and all the pipelines have passed.
- [X] I have made unittests according to the code I have added

### Related issues
Closes #602

### CODE REVIEW

File: hikari/api/event_factory.py [hunk 1]
```diff
@@ -1034,3 +1035,64 @@ def deserialize_voice_server_update_event(
...
+
+    #########################
+    # STAGE INSTANCE EVENTS #
+    #########################
+
+    @abc.abstractmethod
+    def deserialize_stage_instance_create_event(
+        self, shard: gateway_shard.GatewayShard, payload: data_binding.JSONObject
+    ) -> stage_events.StageInstanceCreateEvent:
+        """Parse a raw payload from Discord into a Stage instance create event object.
```
Comment (FasterSpeeding) [inline hikari/api/event_factory.py:1047]: Don't think Stage needs to be capitalised in these descriptions

File: hikari/api/rest.py [hunk 2]
```diff
@@ -7365,3 +7366,185 @@ async def delete_interaction_response(
...
+            The Stage instance associated with the guild stage channel.
+
+            `builtins.None` if no Stage instance exists in the stage channel.
+
+        Raises
+        ------
+        hikari.errors.UnauthorizedError
+            If you are unauthorized to make the request (invalid/missing token).
+        hikari.errors.NotFoundError
+            If the interaction or response is not found.
```
Comment (FasterSpeeding) [inline hikari/api/rest.py:7394]: stage instance not interaction

File: hikari/api/rest.py [hunk 3]
```diff
@@ -7365,3 +7366,185 @@ async def delete_interaction_response(
...
+        ----------
+        channel: hikari.snowflakes.SnowflakeishOr[hikari.channels.GuildStageChannel]
+            The guild stage channel to fetch the Stage instance from.
+
+        Returns
+        -------
+        typing.Optional[hikari.stage_instances.StageInstance]
+            The Stage instance associated with the guild stage channel.
+
+            `builtins.None` if no Stage instance exists in the stage channel.
```
Comment (FasterSpeeding) [inline hikari/api/rest.py:7387]: This doesn't match the type hint

File: hikari/api/rest.py [hunk 4]
```diff
@@ -7365,3 +7366,185 @@ async def delete_interaction_response(
...
+        Parameters
+        ----------
+        channel: hikari.snowflakes.SnowflakeishOr[hikari.channels.GuildStageChannel]
+            The channel to use for the Stage instance creation.
+
+        Other Parameters
+        ----------------
+        topic: builtins.str
+            The topic for the Stage instance.
+
```
Comment (FasterSpeeding) [inline hikari/api/rest.py:7429]: This space between the parameters is unnecessary

File: hikari/api/rest.py [hunk 5]
```diff
@@ -7365,3 +7366,185 @@ async def delete_interaction_response(
...
+        Returns
+        -------
+        hikari.stage_instances.StageInstance
+            The edited Stage instance.
+
+        Raises
+        ------
+        hikari.errors.UnauthorizedError
+            If you are unauthorized to make the request (invalid/missing token).
```
Comment (FasterSpeeding) [inline hikari/api/rest.py:7496]: The permissions behaviour should be mentioned here

File: hikari/api/rest.py [hunk 6]
```diff
@@ -7365,3 +7366,185 @@ async def delete_interaction_response(
...
+        hikari.errors.NotFoundError
+            If the interaction or response is not found.
```
Comment (FasterSpeeding) [inline hikari/api/rest.py:7498]: All these NotFoundError raise reasons need to be updated

File: hikari/events/stage_events.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        Returns
+        -------
+        hikari.snowflakes.Snowflake
+            The ID of the stage instance that this event relates to.
+        """
+
+    @property
+    @abc.abstractmethod
+    def stage_instance(self) -> StageInstance:
+        """Stage Instance that this event relates to.
```
Comment (FasterSpeeding) [inline hikari/events/stage_events.py:64]: the app property should be implemented on StageInstanceEvent

File: hikari/events/stage_events.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+@base_events.requires_intents(intents.Intents.GUILDS)
+class StageInstanceEvent(shard_events.ShardEvent, abc.ABC):
+    """Event base for any event that involves Stage instances."""
+
+    __slots__: typing.Sequence[str] = ()
+
+    @property
+    @abc.abstractmethod
+    def stage_instance_id(self) -> snowflakes.Snowflake:
```
Comment (FasterSpeeding) [inline hikari/events/stage_events.py:52]: instance_id can be implemented here also i think "instance_id" would be a cleaner name

File: hikari/stage_instances.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    guild_id: snowflakes.Snowflake = attr.field(hash=True, repr=False)
+    """The guild ID of the Stage instance."""
+
+    topic: str = attr.field(eq=False, hash=False, repr=False)
+    """The topic of the Stage instance."""
+
+    privacy_level: typing.Union[StagePrivacyLevel, int] = attr.field(eq=False, hash=False, repr=False)
+    """The privacy level of the Stage instance."""
+
+    discoverable_disabled: bool = attr.field(eq=False, hash=False, repr=False)
```
Comment (FasterSpeeding) [inline hikari/stage_instances.py:80]: I assume this field is a moderation thing on discord's end but whatever it is should be clarified in the field's docstring. Also the name of this field should probably be changed to indicate that it's a bool field

File: hikari/stage_instances.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    """Whether or not Stage discovery is disabled."""
+
+    def get_channel(self) -> typing.Optional[channels.GuildStageChannel]:
+        """Return the guild stage channel where this stage instance was created.
+
+        This will be empty if the channels are missing from the cache.
```
Comment (FasterSpeeding) [inline hikari/stage_instances.py:86]: it won't be "empty", it'll return None

File: hikari/stage_instances.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+        Returns
+        -------
+        hikari.channels.GuildStageChannel
+            The guild stage channel where this stage instance was created.
```
Comment (FasterSpeeding) [inline hikari/stage_instances.py:91]: This doesn't match up with the type hint

File: hikari/stage_instances.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+        Returns
+        -------
+        hikari.channels.GuildStageChannel
+            The stage channel where this stage instance was created.
+
+        Raises
+        ------
+        hikari.errors.BadRequestError
+            If any invalid snowflake IDs are passed; a snowflake may be invalid
```
Comment (FasterSpeeding) [inline hikari/stage_instances.py:112]: They aren't passing IDs here

File: hikari/stage_instances.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+        """
+        channel = await self.app.rest.fetch_channel(self.channel_id)
+        assert isinstance(channel, channels.GuildStageChannel)
+
+        return channel
+
+    def get_guild(self) -> typing.Optional[guilds.GatewayGuild]:
+        """Get the cached guild that this Stage Instance relates to, if known.
+
+        If not known, this will return `builtins.None` instead.
```
Comment (FasterSpeeding) [inline hikari/stage_instances.py:142]: better to say "not cache" than "not known"

File: hikari/stage_instances.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        -------
+        hikari.guilds.RESTGuild
+            The guild linked to this Stage Instance
+
+        Raises
+        ------
+        hikari.errors.ForbiddenError
+            If you are not part of the guild.
+        hikari.errors.NotFoundError
+            If the guild is not found.
```
Comment (FasterSpeeding) [inline hikari/stage_instances.py:166]: GET guilds/{guild.id} never returns NotFoundError

**Review by FasterSpeeding — CHANGES_REQUESTED** [review]

**Comment by davfsa** [comment]
Please check comments FasterSpeeding made as well as rebase to master to fix merge conflicts. 

Please also have a look at https://github.com/hikari-py/hikari/blob/master/CONTRIBUTING.md#towncrier on how to use towncrier to document the changes you made

**Comment by FasterSpeeding** [comment]
I'ma take over/finish these PRs next week (I've already talked about this with Ashu)

**Comment by davfsa** [comment]
Closing in favor of #1725

**Pull request closed without merging**